cupy/core/_fusion_trace.pyx

```diff
@@ -0,0 +1,607 @@
+                if arg_id in array_dict:
+                    # The array is already given as an input.
+                    var = in_params[array_dict[arg_id]]
+                    assert isinstance(var, _TraceArray)
+                elif base_id in memory_dict:
+                    # The is an array which shares the same memory.
+                    base = in_params[memory_dict[base_id]]
+                    assert isinstance(base, _TraceArray)
+                    var = self.vc.make_view(base, input_index=input_index)
+                else:
+                    # Otherwise.
+                    var = self.vc.generate_new_array(
+                        arg.dtype, arg.shape, None, input_index=input_index)
+                array_dict[arg_id] = input_index
+                memory_dict[base_id] = input_index
+            else:
+                # Scalar input.
+                dtype = numpy.dtype(type(arg))
+                var = self.vc.generate_new_scalar(
+                    dtype, input_index=input_index)
+            in_params.append(var)
+
+        # Call the target function.
+        inputs = [self._make_interface(x) for x in in_params]
+        output = func(*inputs)
+
+        # Register output variables.
+        if output is None:
+            return_size = 'none'
+            out_params = []
+        elif isinstance(output, _ArrayProxy):
+            return_size = 'single'
+            out_params = [self._unwrap_interface(output, allow_none=True)]
+        elif isinstance(output, tuple):
+            if all(isinstance(x, _ArrayProxy) for x in output):
+                return_size = len(output)
+                out_params = [
+                    self._unwrap_interface(x, allow_none=True) for x in output]
+            else:
+                raise ValueError(
+                    'The all elements of return value of fused function '
+                    'must be of _ArrayProxy type.'
+                )
+        else:
+            raise ValueError(
+                'The return value of fused functions must be `None`, '
+                'ndarray or a tuple of ndarays.'
+            )
+
+        for output_index, out_param in enumerate(out_params):
+            assert isinstance(out_param, _TraceArray)
+            out_param.output_index = output_index
+            out_param.memory.is_output = True
+
+        return out_params, return_size
+
+    def _get_ancestors(self, var):
+        if var is None:
+            return _VariableSet()
+        res = _VariableSet(var)
+        if isinstance(var, _TraceArray):
+            res += self._get_ancestors(var._view_of)
+        return res
+
+    def emit_kernel(self, func, args):
+        # Call `func(args)` and update `op_list`.
+        out_params, return_size = self._trace_target_function(func, args)
+        self.op_list = _fusion_optimization.optimize(
+            self.op_list, self.vc.all_variables, self.shape_constraints)
+
+        # Make info passed to FusedKernel.
```

Comment:
I think it's not `TraceImpl`'s role to prepare kernel parameters. How about returning `out_params` instead?
Perhaps it's better to create a class `_TraceResult` and pass it to `FusedKernel`.